CleanUp

beauremus opened the pull request:
Leverage the features added in #7 like type hinting and test parameterization. Also, clean up flake8 issues.

beauremus committed:
refactor: Use standard tests naming convention.

Also, refer to files using pathlib.

beauremus committed:
style: Enable .vscode linting.

beauremus committed:
refactor: Remove unneeded copy paste.

beauremus committed:
refactor: Use project structure from mCoding.

Using the Python project structure recommended by mCoding.
mCodingLLC/SlapThatLikeButton-TestingStarterProject on GitHub.

beauremus committed:
fix: Remove unneeded and broken coverage report.

beauremus committed:
fix: Refer to correct badge.

I copied the badge from mCoding. It's now referring to our repo.

beauremus committed:
test: Add VSCode pytestEnabled flag.

beauremus committed:
test: Use pytest parameterize. Add docstrings.

beauremus committed:
refactor: Add type hints. Add docstrings.

beauremus committed:
build: Clarify testing process with Make.

Add docs to the README.md file to describe testings process and use
Make to simplify the process.

julianbadillo commented on README.md:
````diff
@@ -124,11 +126,27 @@ To test local modifications, use pip's editable mode.
 
 ```bash
-coverage run -m unittest discover
+make test
````
I have to use `pymake` in Windows. <https://pypi.org/project/py-make/>

beauremus replied:
I didn't think about that. Should we find another tool? I think the standard option is to run commands by hand.

beauremus replied:
Should we instead use a containerized development environment to simplify accounting for various operating system development tools?

julianbadillo replied:
Nah, not worth it, I was reading around and this is pretty standard practice for python development, plus the workflow works fine on Windows. I'm happy if you just make a side note to use `pymake` for Win.

julianbadillo commented on tests/drf_generator_test.py:
```diff
@@ -0,0 +1,116 @@
+    'test_validate2.txt').resolve()
+
+
+def equal(first, second):
```
Wrapping an operator in a function looks like an overkill to me. A lambda might suffice, but then even the operator itself would be enough.

beauremus replied:
We can just pass `==`!? That'd be awesome!

beauremus replied:
I don't think I interpretted what you are saying correctly. What do you mean by "the operator itsefl would be enough"?

julianbadillo commented on tests/drf_generator_test.py:
```diff
@@ -0,0 +1,116 @@
+    """
+    result = list(drf_generator.char_range(first, second))
+    assert comparator(result, expected)
+
```
Why isn't just
```python
assert result == expected
```
I see you'd want to leave room for testing more than one comparator, but it just makes harder to figure out what the test is actually testing, and in failure case, it's a bit convoluted to see the error.
If we want to make another type of list comparison (like inclusion, or permutation), it's ok to write another test case. IMO it's an unneeded abstraction.

julianbadillo reviewed (commented):
Though it's a clever trick, I like the unit tests better without the `comparator` parameter, I had a hard time articulating why beyond _"it looks ugly"_, but these folks sum it up ok: <https://www.nerdwallet.com/blog/engineering/5-pytest-best-practices/>
Bottom line, a test case should test one behavior; **different behaviors should go on different test cases**.
They say something similar here: <https://realpython.com/pytest-python-testing/#parametrization-combining-tests>

> _Even though this shortened your code, it’s important to note that in this case, it didn’t do much to clarify your test code. Use parametrization to separate the test data from the test behavior so that it’s clear what the test is testing!_

beauremus committed:
docs: Reference `py-make` for Windows developers

beauremus committed:
test: Remove comparator arguments.

Per @julianbadillo's suggestion, we should not use the comparator as an
argument because it creates a new logical context and doesn't add
clarity.
